### PR TITLE
 chore: move `primitives/trie` to `reth-trie-types`

### PR DESCRIPTION
**Pull request by joshieDo**
re: https://github.com/paradigmxyz/reth/issues/8674

* turns `primitives/trie` into its own crate.
* moves/adds `primitives/proofs`
* adds and implements trait `IntoTrieAccount` in `primitives/proofs`

Will try to decouple `primitives/proofs` in the next PR

### CODE REVIEW

File: crates/primitives/src/proofs/traits.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+use alloy_genesis::GenesisAccount;
+use alloy_primitives::{keccak256, B256, U256};
+use reth_trie_types::TrieAccount;
+use revm_primitives::AccountInfo;
+
+/// Converts a type into a [`TrieAccount`].
+pub trait IntoTrieAccount {
+    /// Converts to this type into a [`TrieAccount`].
+    fn to_trie_account(self) -> TrieAccount;
+}
```
Comment (mattsse) [inline crates/primitives/src/proofs/traits.rs:12]: hmm, do we even need this?

could we not keep the from impls in trie-types ?

**Review by mattsse — COMMENTED** [review]
> Will try to decouple primitives/proofs in the next PR

sg

question re new conversion trait

please take a look @rkrasiuk

File: crates/primitives/src/proofs/traits.rs (hunk 1 above)
Comment (mattsse) [reply crates/primitives/src/proofs/traits.rs:12]: gotcha, then this is reasonable for now

**Review by mattsse — APPROVED** [review]
lgtm

pending @rkrasiuk

File: crates/trie/types/src/subnode.rs [hunk 2]
```diff
@@ -3,7 +3,7 @@ use bytes::Buf;
 use reth_codecs::Compact;
 
 /// Walker sub node for storing intermediate state root calculation state in the database.
-/// See [`crate::stage::MerkleCheckpoint`].
+/// See `MerkleCheckpoint`.
```
Comment (rkrasiuk) [inline crates/trie/types/src/subnode.rs:6]: no need to reference here
```suggestion
```

File: crates/trie/types/src/account.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    balance: U256,
+    /// Account's storage root.
+    storage_root: B256,
+    /// Hash of the account's bytecode.
+    code_hash: B256,
+}
+
+impl TrieAccount {
+    /// Create new trie account structure.
+    pub const fn new(nonce: u64, balance: U256, storage_root: B256, code_hash: B256) -> Self {
```
Comment (rkrasiuk) [inline crates/trie/types/src/account.rs:19]: this was previously a `From` impl, right? i'd rather remove `new` method and make all fields pub. cautious of mixing up `storage_root` & `code_hash`. e.g. i'd expect `storage_root` to go last

**Review by DaniPopes — APPROVED** [review]

**Review by rkrasiuk — APPROVED** [review]

**Pull request merged**